tracer/src/Datadog.Trace/RemoteConfigurationManagement/RcmPoller.cs

```diff
@@ -0,0 +1,212 @@
+        IRcmApi rcmApi,
+        RcmSettings settings,
+        List<Product> products)
+    {
+        lock (Lock)
+        {
+            if (Instance is null)
+            {
+                var productsMap = products.ToDictionary(p => p.Name, p => p);
+                Instance = new RcmPoller(
+                    discoveryService,
+                    rcmApi,
+                    productsMap,
+                    id: settings.Id,
+                    runtimeId: settings.RuntimeId,
+                    tracerVersion: settings.TracerVersion,
+                    service: settings.ServiceName,
+                    env: settings.Environment,
+                    appVersion: settings.AppVersion,
+                    pollInterval: settings.PollInterval);
+            }
+
+            return Instance;
+        }
+    }
+
+    public async Task StartPollingAsync()
+    {
+        if (string.IsNullOrEmpty(_discoveryService.ConfigurationEndpoint))
+        {
+            Log.Warning("Remote Configuration management is not enabled");
+            return;
+        }
+
+        LifetimeManager.Instance.AddShutdownTask(OnShutdown);
+
+        while (!_cancellationSource.IsCancellationRequested)
+        {
+            try
+            {
+                _lastPollError = null;
+                await Poll().ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                _lastPollError = e.Message;
+            }
+
+            try
+            {
+                await Task.Delay(_pollInterval, _cancellationSource.Token).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // We are shutting down, so don't do anything about it
+            }
+        }
+    }
+
+    private async Task Poll()
+    {
+        var request = BuildRequest();
+        var response = await _rcmApi.GetConfigs(request).ConfigureAwait(false);
+
+        if (response?.Targets?.Signed != null)
+        {
+            ProcessResponse(response);
+            _targetsVersion = response.Targets.Signed.Version;
+        }
+    }
+
+    private GetConfigsRequest BuildRequest()
+    {
+        return new GetConfigsRequest(BuildRequestClient(), new List<CachedTargetFile>());
+
+        RcmClient BuildRequestClient()
+        {
+            return new RcmClient(_id, _products.Keys.ToList(), BuildRcmClientTracer(), BuildRequestClientState());
+        }
+
+        RcmClientTracer BuildRcmClientTracer()
+        {
+            return new RcmClientTracer(_runtimeId, _tracerVersion, _service, _env, _appVersion);
+        }
+
+        RcmClientState BuildRequestClientState()
+        {
+            var configStates =
+                _products
+                   .Values.SelectMany(
+                        product => product.Configs.Select(
+                            config => new RcmConfigState(config.Id, config.Version, config.Product)))
+                   .ToList();
```

Comment:
if `_products` never changes, maybe pre-create the `configStates` once and reuse it for every request?

Reply:
I'm not sure we can assume `_products` will remain static - in the future, both Live Debugger and ASM may be turned off/on remotely through the new "1-click product activation" feature. In those cases, I suppose we will need to enregister/deregister products on the fly? Note that if that's the case, that would also call into question the decision to use C# Events for registering to configuration changes (it may make more sense to explicitly call a method to enregister/deregister like is being [done in Java](https://github.com/DataDog/dd-trace-java/pull/3627))